File: web_version/pmugg_files/shapes3D/edgeType3D.js
ms.edgeType3D = function(faceData, dir, opt_options) {
	// Data consists of {type: faceType3D, onRight: boolean}.
	this.faceData = faceData;
	this.dir = dir;
	
	var options = opt_options || {};

	this.isRigid = options.isRigid || false;
	this.isRigidTiled = options.isRigidTiled || false;
	this.edgeLength = options.edgeLength || Infinity;
	this.angle = options.angle || ms.vec2.angle(ms.vec2.ORIGIN, this.dir);
	this.offset = options.offset || null;
	this.brush = options.brush || null;
	this.monotonic = null;
	// this.is3D = true;
	this.spliced = false;
	// Destroyed means this is part of a super edge and is not part of the gluing options.
	this.destroyed = false;
	
	this.id = ms.counter.add('edgeType3D');
};

ms.counter.register('edgeType3D');

ms.edgeType3D.prototype.export = function(types) {
	return {
		faceData: this.faceData.map((f) => { return { type: types.faceTypes.indexOf(f.type), onRight: f.onRight }}),
		// This might be more complicated for 2D brushes.
		brush: this.brush && this.brush.export && this.brush.export(),

		dir: this.dir.export(),
		isRigid: this.isRigid,
		isRigidTiled: this.isRigidTiled,
		edgeLength: this.edgeLength,
		angle: this.angle,
		offset: this.offset,
		isRigid: this.isRigid,
		spliced: this.spliced,
	};
};

ms.edgeType3D.import = function(json, types) {
	var faceData = json.faceData.map((f) => ({
		type: types.faceTypes[f.type],
		onRight: f.onRight,
	}));	
	var dir = ms.vec3.import(json.dir);
	var options = {...json};
	options.brush = options.brush && (ms.brush.import(5, [options.brush], () => {}));
	var result = new ms.edgeType3D(faceData, dir, options);
	result.setSpliced(json.spliced);
	return result;
};

ms.edgeType3D.partialImport = function(json, faceTypes) {
	var faceData = json.faceData.map((f) => ({
		type: faceTypes[f.type],
		onRight: f.onRight,
	}));	
	var dir = ms.vec3.import(json.dir).swapAxes();	
	var brush = new ms.brush('#000', '#000', () => {})
	var result = new ms.edgeType3D(faceData, dir, {brush: brush});
	result.id = json.id;
	result.idNum = json.idNum;
	return result;
};

ms.edgeType3D.prototype.is3D = function() {
	// return this.is3D;
	// Something is not right.
	return false;
};

ms.edgeType3D.prototype.getDir = function() {
	return this.dir;
};

ms.edgeType3D.prototype.getFaceData = function() {
	return this.faceData;
};

// Angles are not really relevant for 3D edges.
// Use direction when possible.
ms.edgeType3D.prototype.getAngle = function() {
	return this.angle;
};

ms.edgeType3D.prototype.setAngle = function(angle) {
	this.angle = angle;
};

ms.edgeType3D.prototype.setSpliced = function(spliced) {
	this.spliced = spliced;
	if (spliced) {
		this.brush = {getColor: () => '#aaa'};
	}
	return this;
};

ms.edgeType3D.prototype.getSpliced = function() {
	return this.spliced;
};

ms.edgeType3D.prototype.isDestroyed = function() {
	return this.destroyed;
};

ms.edgeType3D.prototype.destroy = function() {
	this.destroyed = true;
};

ms.edgeType3D.prototype.getBrush = function() {
	return this.brush;
};

ms.edgeType3D.prototype.isLoopy = function() {
	return this.brush ? this.brush.get('Loopy') : true;
};

ms.edgeType3D.prototype.isBoundary = function() {
	return this.brush ? this.brush.get('Boundary') : false;
};

ms.edgeType3D.prototype.isConnected = function() {
	if (this.spliced) { return false; }
	return this.brush ? this.brush.get('Fully Connected') : false;
};

ms.edgeType3D.prototype.getThickness = function() {
	return this.brush ? this.brush.get('Thickness') : ms.lineState.INTERSECTION_THICKNESS;
};

// If production rules that split a fragment into multiple fragment is not allowed.
// If there is one connected component, there remains one connected component.
// I may add other single fragments besides the boundary.
ms.edgeType3D.prototype.singleFragment = function() {
	return this.isBoundary();
};

// If production rules that split a fragment into multiple fragment is not allowed.
ms.edgeType3D.prototype.splittable = function() {
	return !(!this.isLoopy() || this.isBoundary() || this.isConnected());
};

ms.edgeType3D.prototype.extendable = function() {
	return !this.isRigid || this.isRigidTiled;
};

ms.edgeType3D.prototype.getEdgeLength = function() {
	return this.edgeLength;
};

ms.edgeType3D.prototype.getIsRigid = function() {
	return this.isRigid;
};

ms.edgeType3D.prototype.getMonotonic = function() {
	return this.monotonic;
};

ms.edgeType3D.prototype.setMonotonic = function(monotonic) {
	this.monotonic = monotonic;
};

ms.edgeType3D.prototype.getId = function() {
	return this.id;
};

ms.edgeType3D.prototype.getLeftArea = function() {
	var leftDatum = this.faceData.find((faceDatum) => { return faceDatum.onRight == false; });
	return leftDatum && leftDatum.type.getMaterial();
};

ms.edgeType3D.prototype.getRightArea = function() {
	var rightDatum = this.faceData.find((faceDatum) => { return faceDatum.onRight == true; });
	return rightDatum && rightDatum.type.getMaterial();
};

// Find the neighboring face that shares a volume with the one at initialIndex.
// above means we are sweeping in the normal direction of initialIndex.
ms.edgeType3D.prototype.neighboringFace = function(initialIndex, above) {
	var maxDim = ms.util.maxDim(this.dir);
	var angles = [];
	this.faceData.map((faceDatum, index) => {
		var x, y;
		var v = this.dir.cross(faceDatum.type.normal);
		if (!faceDatum.onRight) {
			v.scale(-1);
		}
		switch(maxDim) {
			case 0: x = v.x; y = v.y; break;
			case 1: x = v.z; y = v.x; break;
			case 2: x = v.y; y = v.z; break;
		}
		angles.push([Math.atan2(x, y), index]);
	});
	angles.sort((a, b) => { return a[0] > b[0]; });
	var fOrder = angles.findIndex((angle) => angle[1] == initialIndex);
	var neighborOrder = (fOrder + (above ? 1 : -1) + angles.length) % angles.length;
	var neighborIndex = angles[neighborOrder][1];
	return neighborIndex;

	// neighborAbove isn't necessary to calculate. This hasn't been tested.
	/* var initial  = this.faceData[initialIndex];
	var neighbor = this.faceData[neighborIndex]; 
	var n         = initial.type.normal;
	var nNeighbor = neighbor.type.normal;
	var neighborAbove = (n.dot(nNeighbor) < 0) ^ initial.onRight ^ nNeighbor.onRight;
	return {index: neighborIndex, above: neighborAbove}; */
};

/* ms.edgeType3D.prototype.addGraphEndpoints = function(edge, vertex0, vertex1) {
	this.faceData.forEach(function(faceDatum, index) {
		// var vertex = faceDatum.onRight ? vertex0 : vertex1;
		var vertex = vertex0;
		var endpoint = new ms.graphEndpoint(vertex, edge, faceDatum.type);
		edge.setEndpoint(endpoint, index);
		vertex.addEndpoint(endpoint);
	});
}; */

ms.edgeType3D.prototype.boundaryString = function() {
	if (ms.vec3.X_HAT.dot(this.dir) > 0.99) {
		return 'x';
	} else if (ms.vec3.Y_HAT.dot(this.dir) > 0.99) {
		return 'y';
	} else if (ms.vec3.Z_HAT.dot(this.dir) > 0.99) {
		return 'z';
	}
	return String.fromCharCode(this.id + 'a'.charCodeAt(0))
};

// Copied from 2D edgeType. Probably unnecessary
ms.edgeType3D.prototype.addStartVertex = function(v, angle) {
	v.addEdge(this, true, angle);
};

ms.edgeType3D.prototype.addEndVertex = function(v, angle) {
	v.addEdge(this, false, angle);
};

File: cpp_version/pmugg/shapes3D/edge_type3d.cpp
#include "edge_type3d.h"
#include "face_type3d.h"
#include "..\decoration\brush.h"
#include "..\util\util.h"

namespace ms {

int EdgeType3D::nextId = 0;

EdgeType3D::EdgeType3D(const std::vector<FaceData>& fData, const Vec3& direction,
                       const std::map<std::string, bool>& options)
    : faceData(fData)
    , dir(direction)
    , brush(nullptr)
    , angle(std::atan2(dir.getY(), dir.getX()))
    , edgeLength(INFINITY)
    , offset(nullptr)
    , isRigid(options.count("isRigid") ? options.at("isRigid") : false)
    , isRigidTiled(options.count("isRigidTiled") ? options.at("isRigidTiled") : false)
    , monotonic(false)
    , spliced(false)
    , destroyed(false)
    , id(nextId++) {}

void EdgeType3D::setSpliced(bool newSpliced) {
    spliced = newSpliced;
    if (spliced) {
        brush = new Brush("#aaa");
    }
}

bool EdgeType3D::isLoopy() const {
    return brush ? brush->getBool("Loopy") : true;
}

bool EdgeType3D::isBoundary() const {
    return brush ? brush->getBool("Boundary") : false;
}

bool EdgeType3D::isConnected() const {
    if (spliced) return false;
    return brush ? brush->getBool("Fully Connected") : false;
}

bool EdgeType3D::singleFragment() const {
    return isBoundary();
}

bool EdgeType3D::splittable() const {
    return !(!isLoopy() || isBoundary() || isConnected());
}

bool EdgeType3D::extendable() const {
    return !isRigid || isRigidTiled;
}

float EdgeType3D::getThickness() const {
    return brush ? brush->getFloat("Thickness") : 1.0f;
}

std::string EdgeType3D::boundaryString() const {
    if (dir.dot(Vec3::X_AXIS) > 0.99f) return "x";
    if (dir.dot(Vec3::Y_AXIS) > 0.99f) return "y";
    if (dir.dot(Vec3::Z_AXIS) > 0.99f) return "z";
    return std::string(1, static_cast<char>(id + 'a'));
}

int EdgeType3D::neighboringFace(int initialIndex, bool above) const {
    int maxDim = Util::maxDim(dir);
    std::vector<std::pair<float, int>> angles;
    
    for (size_t i = 0; i < faceData.size(); i++) {
        const auto& f = faceData[i];
        Vec3 v = dir.cross(f.type->getNormal());
        if (!f.onRight) {
            v.scale(-1);
        }
        
        float x = 0;
        float y = 0;
        switch(maxDim) {
            case 0: x = v.getX(); y = v.getY(); break;
            case 1: x = v.getZ(); y = v.getX(); break;
            case 2: x = v.getY(); y = v.getZ(); break;
        }
        angles.push_back({std::atan2(x, y), i});
    }
    
    std::sort(angles.begin(), angles.end());
    auto fOrder = std::find_if(angles.begin(), angles.end(),
        [initialIndex](const auto& p) { return p.second == initialIndex; }) - angles.begin();
    
    int neighborOrder = (int)(fOrder + (above ? 1 : -1) + angles.size()) % angles.size();
    return angles[neighborOrder].second;
}

std::string EdgeType3D::getLeftArea() const {
    auto it = std::find_if(faceData.begin(), faceData.end(),
        [](const FaceData& f) { return !f.onRight; });
    return it != faceData.end() ? it->type->getMaterial() : "";
}

std::string EdgeType3D::getRightArea() const {
    auto it = std::find_if(faceData.begin(), faceData.end(),
        [](const FaceData& f) { return f.onRight; });
    return it != faceData.end() ? it->type->getMaterial() : "";
}

/* Json EdgeType3D::export(const Types& types) const {
    Json json;
    
    json["faceData"] = Json::array();
    for (const auto& f : faceData) {
        json["faceData"].push_back({
            {"type", types.faceTypes.indexOf(f.type)},
            {"onRight", f.onRight}
        });
    }
    
    if (brush && brush->export_) {
        json["brush"] = brush->export_();
    }
    
    json["dir"] = dir.export();
    json["isRigid"] = isRigid;
    json["isRigidTiled"] = isRigidTiled;
    json["edgeLength"] = edgeLength;
    json["angle"] = angle;
    json["offset"] = offset ? offset->export() : nullptr;
    json["spliced"] = spliced;
    
    return json;
} */

EdgeType3D* EdgeType3D::import(const Json& json, Shape3D* shape) {
    std::vector<FaceData> fData;
    for (const auto& f : json["faceData"]) {
        fData.push_back({
            shape->faceTypes[f["type"]],
            f["onRight"]
        });
    }
    
    Vec3 direction = Vec3::import(json["dir"]);
    
    std::map<std::string, bool> options = {
        {"isRigid", json["isRigid"]},
        {"isRigidTiled", json["isRigidTiled"]}
    };
    
    auto* result = new EdgeType3D(fData, direction, options);
    
    if (json.contains("brush")) {
        result->brush = Brush::import(json["brush"]);
    }
    // I think edge length is only needed for old grammars that have no brushes.
    // result->edgeLength = json["edgeLength"].is_null() ? Util::INF : json["edgeLength"].get<float>();
    /*result->angle = json.contains("angle") ?
        json["angle"].get<float>() :
        Vec2::angle(Vec2::ORIGIN, result->dir);*/
    if (json["offset"] != nullptr) {
        result->offset = new Vec3(Vec3::import(json["offset"]));
    }
    result->setSpliced(json["spliced"]);
    
    return result;
}

EdgeType3D* EdgeType3D::partialImport(const Json& json, const std::vector<FaceType3D*>& faceTypes) {
    std::vector<FaceData> fData;
    for (const auto& f : json["faceData"]) {
        fData.push_back({
            faceTypes[f["type"]],
            f["onRight"]
        });
    }
    
    Vec3 direction = Vec3::import(json["dir"]).swapAxes();
    auto* brush = new Brush("#000", "#000");
    
    auto* result = new EdgeType3D(fData, direction, {{"brush", brush}});
    result->id = json["id"];
    
    return result;
}

} // namespace ms 